nistats/permutation_tests.py

```diff
@@ -0,0 +1,296 @@
+    smax_parts = np.empty((n_perm_chunk))
+    unc_rank_parts = np.zeros(len(original_stat))
+    # Infer conditions to permute
+    conditions = np.where(con_val != 0.0)
+    t0 = time.time()
+    for i in range(n_perm_chunk):
+        # Shuffle labels of the conditions in the design matrix
+        design_matrices = []
+
+        for frame_times, paradigm, kwargs in design_matrix_objs:
+            perm_paradigm = paradigm.copy()
+            con_idx = np.where(con_val != 0.0)[0]
+            conditions = [con_labels[x] for x in con_idx]
+            selection = np.array([False for x in range(len(paradigm))])
+            for condition in conditions:
+                selection = selection | (perm_paradigm['name'] == condition)
+            sel_labels = perm_paradigm[selection]
+            labels_idx = sel_labels.index.tolist()
+            perm_labels = rng.permutation(sel_labels['name'].tolist())
+            for lidx, perm_label in zip(labels_idx, perm_labels):
+                perm_paradigm.loc[lidx, 'name'] = perm_label
+
+            design_matrices.append(make_design_matrix(frame_times,
+                                                      perm_paradigm,
+                                                      **kwargs))
+
+        # Get permuted stat
+        glm_obj = clone(glm_ref)
+        permuted_stat, = glm_obj.fit(imgs, design_matrices) \
+                                .transform(con_val, contrast_name=con_id,
+                                           output_z=False, output_stat=True)
+        permuted_stat = glm_obj.masker_.transform(permuted_stat)[0]
+
+        # For uncorrected p-values
+        if two_sided_test:
+            unc_rank_parts += (np.fabs(permuted_stat) < original_stat)
+        else:
+            unc_rank_parts += (permuted_stat < original_stat)
+
+        # For corrected p-values
+        smax_parts[i] = np.max(permuted_stat)
+
+        if verbose > 0:
+            # We print every 10 permutations or more often
+            step = 11 - min(verbose, 10)
+            if (i % step == 0):
+                # If there is only one job, progress information is fixed
+                if n_perm == n_perm_chunk:
+                    crlf = "\r"
+                else:
+                    crlf = "\n"
+                percent = float(i) / n_perm_chunk
+                percent = round(percent * 100, 2)
+                dt = time.time() - t0
+                # We use a max to avoid a division by zero
+                remaining = (100. - percent) / max(0.01, percent) * dt
+                sys.stderr.write(
+                    "Job #%d, processed %d/%d permutations for contrast %s "
+                    "(%0.2f%%, %i seconds remaining)%s"
+                    % (thread_id, i, n_perm_chunk, con_id, percent, remaining,
+                       crlf))
+
+    return unc_rank_parts, smax_parts
+
+
+def first_level_permutation_test(contrasts, glm_ref, imgs,
+                                 design_matrix_objs, n_perm=50,
```

Comment:
`n_perm=50` is not  a reasonable default for a number of permutations. I would take 1000